icu4c/source/common/uloc.cpp

```diff
@@ -1199,48 +1165,52 @@ ulocimp_getLanguage(const char *localeID,
 
     /* if it starts with i- or x- then copy that prefix */
     if(_isIDPrefix(localeID)) {
-        if(i<languageCapacity) {
-            language[i]=(char)uprv_tolower(*localeID);
-        }
-        if(i<languageCapacity) {
-            language[i+1]='-';
-        }
-        i+=2;
+        result.append((char)uprv_tolower(*localeID), status);
+        result.append('-', status);
         localeID+=2;
     }
 
     /* copy the language as far as possible and count its length */
     while(!_isTerminator(*localeID) && !_isIDSeparator(*localeID)) {
-        if(i<languageCapacity) {
-            language[i]=(char)uprv_tolower(*localeID);
-        }
-        if(i<3) {
-            U_ASSERT(i>=0);
-            lang[i]=(char)uprv_tolower(*localeID);
-        }
-        i++;
+        result.append((char)uprv_tolower(*localeID), status);
         localeID++;
     }
 
-    if(i==3) {
+    if(result.length()==3) {
         /* convert 3 character code to 2 character code if possible *CWB*/
-        offset=_findIndex(LANGUAGES_3, lang);
+        int32_t offset = _findIndex(LANGUAGES_3, result.data());
         if(offset>=0) {
-            i=_copyCount(language, languageCapacity, LANGUAGES[offset]);
+            result.clear();
+            result.append(LANGUAGES[offset], status);
         }
     }
 
     if(pEnd!=NULL) {
         *pEnd=localeID;
     }
-    return i;
+
+    return result;
 }
 
 U_CFUNC int32_t
+ulocimp_getLanguage(const char *localeID,
+                    char *language, int32_t languageCapacity,
+                    const char **pEnd) {
+    ErrorCode status;
+    CharString result = ulocimp_getLanguage(localeID, pEnd, status);
+    if (status.isFailure()) {
+        return 0;
+    }
+    CheckedArrayByteSink sink(language, languageCapacity);
```

Comment:
This sink does not really buy us anything. You actually have a real bug here -- language[] is not NUL-terminated.

I would
```
int32_t length = result.length();
if (length <= languageCapacity) {
    uprv_memcpy(language, result.data(), length);
    if (length < languageCapacity) { language[length] = 0; }
}
return length;
```

Reply:
> This sink does not really buy us anything.

Done.

Reply:
> You actually have a real bug here -- language[] is not NUL-terminated.

As far as I understand the code this is not a bug but by design, the existing code doesn't `NUL` terminate these buffers and there is no caller of these functions (they're all internal helper functions) that expects that do be done.

The new functions added by this PR, that return `CharString` instead, get automatic `NUL` termination through `CharString` and eventually all code that still calls the old functions will be updated to use the new functions instead, so the issue will eventually simply disappear and the time being, I think that keeping current behaviour intact is the best course of action.